riscv/src/register/macros.rs

```diff
@@ -600,3 +600,510 @@ macro_rules! clear_pmp {
+     $set_field:ident,
+     $(#[$try_set_field_doc:meta])+
+     $try_set_field:ident,
+     range: $bit_start:literal ..= $bit_end:literal$(,)?
+    ) => {
+         $crate::read_only_csr_field!(
+             $ty,
+             $(#[$field_doc])+
+             $field,
+             $(#[$try_field_doc])+
+             $try_field,
+             range: $bit_start ..= $bit_end,
+         );
+
+         $crate::write_only_csr_field!(
+             $ty,
+             $(#[$set_field_doc])+
+             $set_field,
+             $(#[$try_set_field_doc])+
+             $try_set_field,
+             range: $bit_start ..= $bit_end,
+         );
+    };
+
+    ($ty:ident,
+     $(#[$field_doc:meta])+
+     $field:ident,
+     $(#[$set_field_doc:meta])+
+     $set_field:ident,
+     range: [$bit_start:literal : $bit_end:literal]$(,)?
+    ) => {
+        $crate::read_only_csr_field!(
+            $ty,
+            $(#[$field_doc])+
+            $field: [$bit_start : $bit_end],
+        );
+
+        $crate::write_only_csr_field!(
+            $ty,
+            $(#[$set_field_doc])+
+            $set_field: [$bit_start : $bit_end],
+        );
+    };
+
+    ($ty:ident,
+     $(#[$field_doc:meta])+
+     $field:ident,
+     $(#[$try_field_doc:meta])+
+     $try_field:ident,
+     $(#[$set_field_doc:meta])+
+     $set_field:ident,
+     $(#[$field_ty_doc:meta])+
+     $field_ty:ident {
+         range: [$field_start:literal : $field_end:literal],
+         default: $default_variant:ident,
+         $($variant:ident = $value:expr$(,)?)+
+     }$(,)?
+    ) => {
+        $crate::csr_field_enum!(
+            $(#[$field_ty_doc])+
+            $field_ty {
+                range: [$field_start : $field_end],
+                default: $default_variant,
+                $($variant = $value,)+
+            },
+         );
+
+         $crate::read_only_csr_field!(
+             $ty,
+             $(#[$field_doc])+
+             $field,
+             $(#[$try_field_doc])+
+             $try_field,
+             $field_ty,
+             range: [$field_start : $field_end],
+         );
+
+         $crate::write_only_csr_field!(
+             $ty,
+             $(#[$set_field_doc])+
+             $set_field,
+             $field_ty,
+             range: [$field_start : $field_end],
+         );
+    };
+}
+
+/// Defines field accesor functions for a read-only CSR type.
+#[macro_export]
+macro_rules! read_only_csr_field {
+    ($ty:ident,
+     $(#[$field_doc:meta])+
+     $field:ident: $bit:literal$(,)?) => {
+        const _: () = assert!($bit < usize::BITS);
+
+        impl $ty {
+            $(#[$field_doc])+
+            #[inline]
+            pub fn $field(&self) -> bool {
+                $crate::bits::bf_extract(self.bits, $bit, 1) != 0
+            }
+        }
+    };
+
+    ($ty:ident,
+     $(#[$field_doc:meta])+
+     $field:ident,
+     $(#[$try_field_doc:meta])+
+     $try_field:ident,
+     range: $bit_start:literal..=$bit_end:literal$(,)?) => {
```

Comment:
I would use to `start : end` for consistency with the other branches
```suggestion
     range: $bit_start:literal : $bit_end:literal$(,)?) => {
```

Reply:
This is specifically to handle a range of single-bit fields, like `hpm` in `mcountinhibit`.

The `[start : end]` variant is for a multi-bit field.